webapi/src/main/scala/org/knora/webapi/store/cacheservice/impl/CacheServiceInMemImpl.scala

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright © 2021 - 2022 Swiss National Data and Service Center for the Humanities and/or DaSCH Service Platform contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.knora.webapi.store.cacheservice.impl
+
+import akka.http.scaladsl.util.FastFuture
+import org.knora.webapi.IRI
+import org.knora.webapi.messages.admin.responder.projectsmessages.{
+  ProjectADM,
+  ProjectIdentifierADM,
+  ProjectIdentifierType
+}
+import org.knora.webapi.messages.admin.responder.usersmessages.{UserADM, UserIdentifierADM, UserIdentifierType}
+import org.knora.webapi.messages.store.cacheservicemessages.{CacheServiceStatusOK, CacheServiceStatusResponse}
+import org.knora.webapi.store.cacheservice.api.{CacheService, EmptyKey, EmptyValue}
+import zio._
+import zio.stm._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+ * In-Memory Cache implementation
+ *
+ * The state is divided into Refs used to store different types of objects.
+ * A ref in itself is fiber (thread) safe, but to keep the cumulative state
+ * consistent, all Refs need to be updated in a single transaction. This
+ * requires STM (Software Transactional Memory) to be used.
+ *
+ * @param users a map of users.
+ * @param projects a map of projects.
+ * @param lut   a lookup table of username/email to IRI.
+ */
+case class CacheServiceInMemImpl(
+  users: TMap[String, UserADM],
+  projects: TMap[String, ProjectADM],
+  lut: TMap[String, String] // sealed trait for key type
+) extends CacheService {
+
+  /**
+   * Stores the user under the IRI (inside 'users') and additionally the IRI
+   * under the keys of USERNAME and EMAIL (inside the 'lut'):
+   *
+   * IRI -> byte array
+   * username -> IRI
+   * email -> IRI
+   *
+   * @param value the value to be stored
+   */
+  def putUserADM(value: UserADM): Task[Unit] =
+    (for {
+      _ <- users.put(value.id, value)
+      _ <- lut.put(value.username, value.id)
+      _ <- lut.put(value.email, value.id)
+    } yield ()).commit.tap(_ => ZIO.logInfo(s"Stored UserADM to Cache: ${value.id}"))
+
+  /**
+   * Retrieves the user stored under the identifier (either iri, username, or email).
+   *
+   * The data is stored under the IRI key.
+   * Additionally, the USERNAME and EMAIL keys point to the IRI key
+   *
+   * @param identifier the user identifier.
+   */
+  def getUserADM(identifier: UserIdentifierADM): Task[Option[UserADM]] =
+    (identifier.hasType match {
+      case UserIdentifierType.Iri      => getUserByIri(identifier.toIri)
+      case UserIdentifierType.Username => getUserByUsernameOrEmail(identifier.toUsername)
+      case UserIdentifierType.Email    => getUserByUsernameOrEmail(identifier.toEmail)
+    }).tap(_ => ZIO.logInfo(s"Retrieved UserADM from Cache: ${identifier}"))
+
+  /**
+   * Retrieves the user stored under the IRI.
+   *
+   * @param id the user's IRI.
+   * @return an optional [[UserADM]].
```

Comment:
I'm still not sure what the best way would be for writing docstings. But I didn't invest any time on figuring it out :-) Probably looking at the ZIO codebase will give some insights. I wrote optional there, because `UserADM` is the important piece of information. In the editor you will see now the complete type. With Akka this was not the case.